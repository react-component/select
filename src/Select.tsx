/**
 * To match accessibility requirement, we always provide an input in the component.
 * Other element will not set `tabIndex` to avoid `onBlur` sequence problem.
 * For focused select, we set `aria-live="polite"` to update the accessibility content.
 *
 * ref:
 * - keyboard: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role#Keyboard_interactions
 *
 * New api:
 * - listHeight
 * - listItemHeight
 * - component
 *
 * Remove deprecated api:
 * - multiple
 * - tags
 * - combobox
 * - firstActiveValue
 * - dropdownMenuStyle
 * - openClassName (Not list in api)
 *
 * Update:
 * - `backfill` only support `combobox` mode
 * - `combobox` mode not support `labelInValue` since it's meaningless
 * - `getInputElement` only support `combobox` mode
 * - `onChange` return OptionData instead of ReactNode
 * - `filterOption` `onChange` `onSelect` accept OptionData instead of ReactNode
 * - `combobox` mode trigger `onChange` will get `undefined` if no `value` match in Option
 * - `combobox` mode not support `optionLabelProp`
 */

import * as React from 'react';
import { OptionsType as SelectOptionsType } from './interface';
import SelectOptionList from './OptionList';
import Option from './Option';
import OptGroup from './OptGroup';
import { convertChildrenToData as convertSelectChildrenToData } from './utils/legacyUtil';
import {
  getLabeledValue as getSelectLabeledValue,
  filterOptions as selectDefaultFilterOptions,
  isValueDisabled as isSelectValueDisabled,
  findValueOption as findSelectValueOption,
  flattenOptions,
  fillOptionsWithMissingValue,
} from './utils/valueUtil';
import generateSelector, { SelectProps, RefSelectProps } from './generate';
import { DefaultValueType, Key, SingleType } from './interface/generator';
import warningProps from './utils/warningPropsUtil';

const RefSelect = generateSelector<SelectOptionsType>({
  prefixCls: 'rc-select',
  components: {
    optionList: SelectOptionList,
  },
  convertChildrenToData: convertSelectChildrenToData,
  flattenOptions,
  getLabeledValue: getSelectLabeledValue,
  filterOptions: selectDefaultFilterOptions,
  isValueDisabled: isSelectValueDisabled,
  findValueOption: findSelectValueOption,
  warningProps,
  fillOptionsWithMissingValue,
});

export type ExportedSelectProps<
  ValueType extends DefaultValueType = DefaultValueType
> = SelectProps<SelectOptionsType, ValueType>;

/**
 * Typescript not support generic with function component,
 * we have to wrap an class component to handle this.
 */
class Select<
  VT extends Key | Key[] = Key,
  TKeyForOptions extends Key = Key
> extends React.Component<SelectProps<SelectOptionsType<TKeyForOptions, SingleType<VT>>, VT>> {
  static Option: typeof Option = Option;

  static OptGroup: typeof OptGroup = OptGroup;

  selectRef = React.createRef<RefSelectProps>();

  focus = () => {
    this.selectRef.current.focus();
  };

  blur = () => {
    this.selectRef.current.blur();
  };

  render() {
    return <RefSelect ref={this.selectRef} {...this.props} />;
  }
}

export default Select;
