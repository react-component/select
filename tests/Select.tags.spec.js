/* eslint-disable no-undef */
import React from 'react';
import { mount, render } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import Select, { Option, OptGroup } from '../src';
import allowClearTest from './shared/allowClearTest';
import focusTest from './shared/focusTest';
import blurTest from './shared/blurTest';
import hoverTest from './shared/hoverTest';
import renderTest from './shared/renderTest';
import removeSelectedTest from './shared/removeSelectedTest';
import throwOptionValue from './shared/throwOptionValue';
import dynamicChildrenTest from './shared/dynamicChildrenTest';
import inputFilterTest from './shared/inputFilterTest';
import openControlledTest from './shared/openControlledTest';

describe('Select.tags', () => {
  allowClearTest('tags');
  focusTest('tags');
  blurTest('tags');
  hoverTest('tags');
  renderTest('tags');
  removeSelectedTest('tags');
  throwOptionValue('tags');
  dynamicChildrenTest('tags');
  inputFilterTest('tags');
  openControlledTest('tags');

  it('allow user input tags', () => {
    const wrapper = mount(<Select tags />);

    wrapper
      .find('input')
      .simulate('change', { target: { value: 'foo' } })
      .simulate('keyDown', { keyCode: KeyCode.ENTER });

    expect(wrapper.state().value).toEqual(['foo']);
    expect(
      wrapper
        .update()
        .find('.rc-select-selection__choice__content')
        .text(),
    ).toBe('foo');
  });

  it('should call onChange on blur', () => {
    const wrapper = mount(<Select tags />);

    jest.useFakeTimers();
    wrapper
      .find('input')
      .simulate('change', { target: { value: 'foo' } })
      .simulate('blur');

    jest.runAllTimers();
    expect(wrapper.state().value).toEqual(['foo']);
    expect(
      wrapper
        .update()
        .find('.rc-select-selection__choice__content')
        .text(),
    ).toBe('foo');
  });

  it('tokenize input', () => {
    const handleChange = jest.fn();
    const handleSelect = jest.fn();
    const option2 = <Option value="2">2</Option>;
    const wrapper = mount(
      <Select tags tokenSeparators={[',']} onChange={handleChange} onSelect={handleSelect}>
        <Option value="1">1</Option>
        {option2}
      </Select>,
    );

    const input = wrapper.find('input');
    input.instance().focus = jest.fn();

    input.simulate('change', { target: { value: '2,3,4' } });

    expect(handleChange).toBeCalledWith(['2', '3', '4'], expect.anything());
    expect(handleSelect).toHaveBeenCalledTimes(3);
    expect(handleSelect).toHaveBeenLastCalledWith(
      '4',
      <Option key="4" value="4">
        4
      </Option>,
    );
    expect(wrapper.state().value).toEqual(['2', '3', '4']);
    expect(
      wrapper
        .find('.rc-select-selection__choice__content')
        .at(0)
        .text(),
    ).toBe('2');
    expect(
      wrapper
        .find('.rc-select-selection__choice__content')
        .at(1)
        .text(),
    ).toBe('3');
    expect(
      wrapper
        .find('.rc-select-selection__choice__content')
        .at(2)
        .text(),
    ).toBe('4');
    expect(wrapper.state().inputValue).toBe('');
    expect(wrapper.state().open).toBe(false);
    expect(input.instance().focus).toBeCalled();
  });

  it('renders unlisted item in value', () => {
    const wrapper = render(
      <Select tags value="3" open>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('renders search value when not found', () => {
    const wrapper = render(
      <Select tags value="22" inputValue="2" open>
        <Option value="1">1</Option>
      </Select>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('use filterOption', () => {
    const filterOption = (inputValue, option) =>
      option.props.value.toLowerCase().indexOf(inputValue.toLowerCase()) !== -1;

    const wrapper = render(
      <Select tags inputValue="red" filterOption={filterOption} open>
        <Option value="Red">Red</Option>
      </Select>,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('filterOption is false', () => {
    const wrapper = mount(
      <Select tags filterOption={false}>
        <Option value="1">1</Option>
        <Option value="2">2</Option>
      </Select>,
    );
    const input = wrapper.find('input');
    input.instance().focus = jest.fn();
    input
      .simulate('change', { target: { value: 'a' } })
      .simulate('keyDown', { keyCode: KeyCode.ENTER });

    expect(wrapper.state().value).toEqual(['a']);
    expect(wrapper.find('.rc-select-selection__choice__content').text()).toBe('a');
  });

  describe('OptGroup', () => {
    const createSelect = props => (
      <div>
        <Select tags {...props}>
          <OptGroup key="Manager" label="Manager">
            <Option key="jack" value="jack">
              Jack
            </Option>
          </OptGroup>
          <OptGroup key="Engineer" label="Engineer">
            <Option key="Yiminghe" value="Yiminghe">
              yiminghe
            </Option>
          </OptGroup>
        </Select>
      </div>
    );

    it('renders correctly', () => {
      const wrapper = mount(createSelect({ value: ['jack', 'foo'] }));
      wrapper.find('.rc-select').simulate('click');
      expect(wrapper.render()).toMatchSnapshot();
    });

    it('renders inputValue correctly', () => {
      const wrapper = mount(createSelect());
      wrapper.find('.rc-select').simulate('click');

      wrapper.find('input').simulate('change', { target: { value: 'foo' } });
      expect(wrapper.render()).toMatchSnapshot();

      wrapper.find('input').simulate('keyDown', { keyCode: KeyCode.ENTER });
      expect(wrapper.render()).toMatchSnapshot();
    });
  });
});
