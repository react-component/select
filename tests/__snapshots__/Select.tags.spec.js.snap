// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select.tags allowClear renders correctly 1`] = `
<div
  class="rc-select rc-select-enabled rc-select-allow-clear"
>
  <div
    aria-autocomplete="list"
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.tags render render animation 1`] = `
<SelectTrigger
  defaultActiveFirstOption={true}
  dropdownMatchSelectWidth={true}
  dropdownMenuStyle={Object {}}
  dropdownStyle={Object {}}
  inputValue=""
  multiple={true}
  onDropdownVisibleChange={[Function]}
  onMenuDeselect={[Function]}
  onMenuSelect={[Function]}
  onPopupFocus={[Function]}
  options={Array []}
  prefixCls="rc-select"
  showAction={
    Array [
      "click",
    ]
  }
  showSearch={true}
  value={Array []}
  visible={false}
>
  <div
    className="rc-select rc-select-enabled"
    onBlur={[Function]}
    onFocus={[Function]}
  >
    <div
      aria-autocomplete="list"
      aria-expanded={false}
      aria-haspopup="true"
      className="rc-select-selection
            rc-select-selection--multiple"
      key="selection"
      role="combobox"
    >
      <div
        className="rc-select-selection__rendered"
      >
        <Animate
          animation={Object {}}
          component="ul"
          componentProps={Object {}}
          onAppear={[Function]}
          onEnd={[Function]}
          onEnter={[Function]}
          onLeave={[Function]}
          transitionAppear={false}
          transitionEnter={true}
          transitionLeave={true}
          transitionName="slide-up"
        >
          <li
            className="rc-select-search rc-select-search--inline"
            key="__input"
          >
            <div
              className="rc-select-search__field__wrap"
            >
              <input
                autoComplete="off"
                className="rc-select-search__field"
                onChange={[Function]}
                onKeyDown={[Function]}
                value=""
              />
              <span
                className="rc-select-search__field__mirror"
              >
                 
              </span>
            </div>
          </li>
        </Animate>
      </div>
    </div>
  </div>
</SelectTrigger>
`;

exports[`Select.tags render truncates tags by maxTagCount 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice"
          style="user-select:none;-webkit-user-select:none"
          title="one"
          unselectable="unselectable"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            one
          </div>
          <span
            class="rc-select-selection__choice__remove"
          />
        </li>
        <li
          class="rc-select-selection__choice"
          style="user-select:none;-webkit-user-select:none"
          title="two"
          unselectable="unselectable"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            two
          </div>
          <span
            class="rc-select-selection__choice__remove"
          />
        </li>
        <li
          class="rc-select-selection__choice rc-select-selection__choice__disabled"
          style="user-select:none;-webkit-user-select:none"
          title="[object Object]"
          unselectable="unselectable"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            + 1 ...
          </div>
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.tags render truncates values by maxTagTextLength 1`] = `
<div
  class="rc-select rc-select-enabled"
>
  <div
    aria-autocomplete="list"
    aria-expanded="false"
    aria-haspopup="true"
    class="rc-select-selection
            rc-select-selection--multiple"
    role="combobox"
  >
    <div
      class="rc-select-selection__rendered"
    >
      <ul>
        <li
          class="rc-select-selection__choice"
          style="user-select:none;-webkit-user-select:none"
          title="one"
          unselectable="unselectable"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            on...
          </div>
          <span
            class="rc-select-selection__choice__remove"
          />
        </li>
        <li
          class="rc-select-selection__choice"
          style="user-select:none;-webkit-user-select:none"
          title="two"
          unselectable="unselectable"
        >
          <div
            class="rc-select-selection__choice__content"
          >
            tw...
          </div>
          <span
            class="rc-select-selection__choice__remove"
          />
        </li>
        <li
          class="rc-select-search rc-select-search--inline"
        >
          <div
            class="rc-select-search__field__wrap"
          >
            <input
              autocomplete="off"
              class="rc-select-search__field"
              value=""
            />
            <span
              class="rc-select-search__field__mirror"
            >
               
            </span>
          </div>
        </li>
      </ul>
    </div>
  </div>
</div>
`;

exports[`Select.tags renders search value when not found 1`] = `
Array [
  <div
    class="rc-select rc-select-open rc-select-focused rc-select-enabled"
  >
    <div
      aria-autocomplete="list"
      aria-expanded="true"
      aria-haspopup="true"
      class="rc-select-selection
            rc-select-selection--multiple"
      role="combobox"
    >
      <div
        class="rc-select-selection__rendered"
      >
        <ul>
          <li
            class="rc-select-selection__choice"
            style="user-select:none;-webkit-user-select:none"
            title="22"
            unselectable="unselectable"
          >
            <div
              class="rc-select-selection__choice__content"
            >
              22
            </div>
            <span
              class="rc-select-selection__choice__remove"
            />
          </li>
          <li
            class="rc-select-search rc-select-search--inline"
          >
            <div
              class="rc-select-search__field__wrap"
            >
              <input
                autocomplete="off"
                class="rc-select-search__field"
                value=""
              />
              <span
                class="rc-select-search__field__mirror"
              >
                 
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  <div>
    <div
      class="rc-select-dropdown rc-select-dropdown--multiple rc-select-dropdown-placement-bottomLeft "
    >
      <div
        style="overflow:auto"
      >
        <ul
          aria-activedescendant=""
          class="rc-select-dropdown-menu  rc-select-dropdown-menu-root rc-select-dropdown-menu-vertical"
          role="menu"
        >
          <li
            aria-selected="false"
            class="rc-select-dropdown-menu-item"
            role="menuitem"
            style="user-select:none;-webkit-user-select:none"
            unselectable="unselectable"
          >
            1
          </li>
          <li
            aria-selected="true"
            class="rc-select-dropdown-menu-item rc-select-dropdown-menu-item-active rc-select-dropdown-menu-item-selected"
            role="menuitem"
            style="user-select:none;-webkit-user-select:none"
            unselectable="unselectable"
          >
            22
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;

exports[`Select.tags renders unlisted item in value 1`] = `
Array [
  <div
    class="rc-select rc-select-open rc-select-focused rc-select-enabled"
  >
    <div
      aria-autocomplete="list"
      aria-expanded="true"
      aria-haspopup="true"
      class="rc-select-selection
            rc-select-selection--multiple"
      role="combobox"
    >
      <div
        class="rc-select-selection__rendered"
      >
        <ul>
          <li
            class="rc-select-selection__choice"
            style="user-select:none;-webkit-user-select:none"
            title="3"
            unselectable="unselectable"
          >
            <div
              class="rc-select-selection__choice__content"
            >
              3
            </div>
            <span
              class="rc-select-selection__choice__remove"
            />
          </li>
          <li
            class="rc-select-search rc-select-search--inline"
          >
            <div
              class="rc-select-search__field__wrap"
            >
              <input
                autocomplete="off"
                class="rc-select-search__field"
                value=""
              />
              <span
                class="rc-select-search__field__mirror"
              >
                 
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  <div>
    <div
      class="rc-select-dropdown rc-select-dropdown--multiple rc-select-dropdown-placement-bottomLeft "
    >
      <div
        style="overflow:auto"
      >
        <ul
          aria-activedescendant=""
          class="rc-select-dropdown-menu  rc-select-dropdown-menu-root rc-select-dropdown-menu-vertical"
          role="menu"
        >
          <li
            aria-selected="false"
            class="rc-select-dropdown-menu-item"
            role="menuitem"
            style="user-select:none;-webkit-user-select:none"
            unselectable="unselectable"
          >
            1
          </li>
          <li
            aria-selected="false"
            class="rc-select-dropdown-menu-item"
            role="menuitem"
            style="user-select:none;-webkit-user-select:none"
            unselectable="unselectable"
          >
            2
          </li>
          <li
            aria-selected="true"
            class="rc-select-dropdown-menu-item rc-select-dropdown-menu-item-active rc-select-dropdown-menu-item-selected"
            role="menuitem"
            style="user-select:none;-webkit-user-select:none"
            unselectable="unselectable"
          >
            3
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;

exports[`Select.tags use filterOption 1`] = `
Array [
  <div
    class="rc-select rc-select-open rc-select-focused rc-select-enabled"
  >
    <div
      aria-autocomplete="list"
      aria-expanded="true"
      aria-haspopup="true"
      class="rc-select-selection
            rc-select-selection--multiple"
      role="combobox"
    >
      <div
        class="rc-select-selection__rendered"
      >
        <ul>
          <li
            class="rc-select-search rc-select-search--inline"
          >
            <div
              class="rc-select-search__field__wrap"
            >
              <input
                autocomplete="off"
                class="rc-select-search__field"
                value=""
              />
              <span
                class="rc-select-search__field__mirror"
              >
                 
              </span>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>,
  <div>
    <div
      class="rc-select-dropdown rc-select-dropdown--multiple rc-select-dropdown-placement-bottomLeft "
    >
      <div
        style="overflow:auto"
      >
        <ul
          aria-activedescendant=""
          class="rc-select-dropdown-menu  rc-select-dropdown-menu-root rc-select-dropdown-menu-vertical"
          role="menu"
        >
          <li
            aria-selected="false"
            class="rc-select-dropdown-menu-item rc-select-dropdown-menu-item-active"
            role="menuitem"
            style="user-select:none;-webkit-user-select:none"
            unselectable="unselectable"
          >
            Red
          </li>
        </ul>
      </div>
    </div>
  </div>,
]
`;
